samples/nrf9160/lwm2m_client/src/main.c

```diff
@@ -306,11 +307,27 @@ void main(void)
 
 #if !defined(CONFIG_NRF_MODEM_LIB_SYS_INIT)
 	ret = nrf_modem_lib_init(NORMAL_MODE);
-	if (ret != 0 && ret != MODEM_DFU_RESULT_OK) {
+#else
+	ret = nrf_modem_lib_get_init_ret();
+#endif
+	if (ret != 0) {
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT)
+		/* Modem FW update needs to be verified before modem is used. */
+		lwm2m_verify_modem_fw_update(ret);
+#else
 		LOG_ERR("Unable to init modem library (%d)", ret);
 		return;
+#endif
+		}
+
+#if !defined(CONFIG_NRF_MODEM_LIB_SYS_INIT) && defined(CONFIG_PDN)
+	ret = pdn_init();
+	if (ret < 0) {
+		LOG_ERR("Unable to init pdn (%d)", ret);
+		return;
 	}
 #endif
+
```

Comment:
I think this looks way too complex now.
Can you refactor this a bit

```c
#if !defined(CONFIG_NRF_MODEM_LIB_SYS_INIT)
        ret = nrf_modem_lib_init(NORMAL_MODE);
        if (ret != 0) {
            LOG_ERR("Unable to init modem library (%d)", ret);
	    return;
        }
#endif

#if defined(CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT)
	/* Modem FW update needs to be verified before modem is used. */
	lwm2m_verify_modem_fw_update();
#endif

#if defined(CONFIG_PDN)
	ret = pdn_init();
	if (ret < 0) {
		LOG_ERR("Unable to init pdn (%d)", ret);
		return;
	}
#endif
```

Reply:
Call nrf_modem_lib_init() in app could return MODEM_DFU_RESULT_xxxx (just like SYS_INIT calls _nrf_modem_lib_init) that must be handled by lwm2m_verify_modem_fw_update(), so cannot return by if (ret != 0)

Reply:
Ah, you are right..
This `ret != 0` could be changed to `ret < 0` because
```
/* * @retval Zero on success.
 * @retval A positive value from @ref nrf_modem_dfu when executing
 *         Modem firmware updates.
 *
 * @retval -NRF_EFAULT @c init_params is @c NULL.
 * @retval -NRF_ENOMEM Not enough shared memory for this operation.
 * @retval -NRF_EPERM The Modem library is already initialized.
 * @retval -NRF_ETIMEDOUT Modem timed out.
 * @retval -NRF_EIO IPC State fault or missing root digest.
 */
int nrf_modem_init(const nrf_modem_init_params_t *init_params,
		   enum nrf_modem_mode_t mode);
```
So positive values are DFU return codes.